kafka-message-consumer-impl/src/main/java/com/example/consumer/impl/KafkaMessageConsumer.java

```diff
@@ -1,37 +1,57 @@
 package com.example.consumer.impl;
 
-import akka.Done;
 import akka.stream.javadsl.Flow;
+import com.example.consumer.impl.entity.DummyEntity;
+import com.example.consumer.impl.entity.MessageCommand;
 import com.example.generator.api.KafkaMessageGeneratorService;
+import com.example.generator.api.Message;
+import com.lightbend.lagom.javadsl.persistence.PersistentEntityRegistry;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import scala.concurrent.duration.FiniteDuration;
 
 import javax.inject.Inject;
 import javax.inject.Singleton;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 
 @Singleton
 public class KafkaMessageConsumer {
     private final Logger log = LoggerFactory.getLogger(getClass());
     private final AtomicBoolean started = new AtomicBoolean(false);
     private final KafkaMessageGeneratorService kafkaMessageGeneratorService;
+    private PersistentEntityRegistry entityRegistry;
+
 
     @Inject
-    public KafkaMessageConsumer(KafkaMessageGeneratorService kafkaMessageGeneratorService) {
+    public KafkaMessageConsumer(
+            KafkaMessageGeneratorService kafkaMessageGeneratorService,
+            PersistentEntityRegistry entityRegistry) {
+
         this.kafkaMessageGeneratorService = kafkaMessageGeneratorService;
+        this.entityRegistry = entityRegistry;
+
+        entityRegistry.register(DummyEntity.class);
     }
 
+
     void start() {
-        if (started.compareAndSet(false, true)) {
-            kafkaMessageGeneratorService.messages()
-                    .subscribe()
-                    .atLeastOnce(
-                            Flow.fromFunction(m -> {
-                                log.info("Consumed message: " + m);
-                                return Done.getInstance();
-                            })
-                    );
-
-        }
+        kafkaMessageGeneratorService.messages()
+                .subscribe()
+                .atLeastOnce(
+                        Flow.<Message>create()
+                                .groupedWithin(3,  new FiniteDuration(1, TimeUnit.SECONDS))
+                                .mapAsync(
+                                        1,
+                                        (List<Message> m) -> {
+                                            log.info("Consumed message: " + m);
+                                            return entityRegistry
+                                                    .refFor(DummyEntity.class, m.get(0).getId())
+                                                    .ask(new MessageCommand(m.get(0).getMessage()));
+                                        })
```

Comment:
The offset mismatch you mentioned over email is happening because this returns one `Done` for each batch.

I think what you would want to do for a realistic simulation is map the list of messages to a list of `Done`s and then concat that into the stream.

Of course, this would make the batching redundant, so an even better example would be one where the entity command itself takes a list. Either it would return an equivalently-sized list of results, or you would need to unfold the result into a `Done` for each input element.

Reply:
OK I see you have handled this in #2